doc/neps/nep-0040-legacy-datatype-impl.rst

```diff
@@ -0,0 +1,616 @@
+================================================
+NEP 40 — Legacy Datatype Implementation in NumPy
+================================================
+
+:title: Legacy Datatype Implementation in NumPy
+:Author: Sebastian Berg
+:Status: Draft
+:Type: Informational
+:Created: 2019-07-17
+
+
+Abstract
+--------
+
+As a preparation to further NumPy enhancement proposals 41, 42, and 43. This
+NEP details the current status of NumPy datatypes as of NumPy 1.18.
+It describes some of the technical aspects and concepts that
+motivated the other proposals.
+
+
+Detailed Description
+--------------------
+
+This section describes some central concepts and provides a brief overview
+of the current implementation of dtypes as well as a discussion.
+In many cases subsections will be split roughly to first describe the
+current implementation and then follow with an "Issues and Discussion" section.
+
+Parametric Datatypes
+^^^^^^^^^^^^^^^^^^^^
+
+Some datatypes are inherently *parametric*. All ``np.flexible`` scalar
+types are attached to parametric datatypes (string, bytes, and void).
+The class ``np.flexible`` for scalars is a superclass for the data types of
+variable length (string, bytes,
+and void). This distinction is similarly exposed by the C-Macros
+``PyDataType_ISFLEXIBLE`` and ``PyTypeNum_ISFLEXIBLE``.
+For instance, ``"S8"`` can represent longer strings than ``"S4"``.
+
+The basic numerical datatypes are not flexible (do not inherit from
+``npflexible``). Float64, Float32, etc. do have a byte order, but the described
+values are unaffected by it, and it is always possible to cast them to the
+native, canonical representation without any loss of information.
+
+The concept of flexibility can be generalized to parametric datatypes.
+For example the private ``AdaptFlexibleDType`` function also accepts the
+naive datetime dtype as input to find the correct time unit.
+The datetime dtype is thus parametric not in the storage, but instead in
+what the stored value represents.
+Currently ``np.can_cast("datetime64[s]", "datetime64[ms]", casting="safe")``
+returns true, although it is unclear that this is desired or generalizes
+to possible future data types such as physical units.
+
+Thus we have data types (mainly strings) with the properties that:
+
+1. Casting is not always safe (``np.can_cast("S8", "S4")``)
+2. Array coercion should be able to discover the exact dtype, such as for
+   ``np.array(["str1", 123.], dtype="S")`` where NumPy discovers the
+   resulting dtype as ``"S5"``.
+   (Without ``dtype="S"`` such behaviour is currently ill defined [gh-15327]_.)
```

Comment:
Do you mean "if the dtype argument is ommitted"? The behavior is not ill-defined if `np.object_` is passed, for instance.